File: 100. 相同的树/ac.cpp
#include<stack>
#include<iostream>
using namespace std;



struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
 
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        stack<TreeNode> left, right;
        while (1)
        {
            if(p && q && p->val == q->val){
                left.push(*p);
                right.push(*q);
                p = p->left;
                q = q->left;
            }else if (p == nullptr && q == nullptr)
            {
                if(left.empty() && right.empty())
                    return true;
                p = &left.top();
                p = p->right;
                q = &right.top();
                q = q->right;
                left.pop();
                right.pop();

            }else{
                return false;
            }
            
        }
        
    }
};

int main(){
    Solution s;
    TreeNode *a = new TreeNode(1);
    TreeNode *b = new TreeNode(2);
    TreeNode *c = new TreeNode(3);
    TreeNode *x = new TreeNode(1);
    TreeNode *y = new TreeNode(2);
    TreeNode *z = new TreeNode(5);
    a->left = b;
    a->right = c;
    x->left = y;
    x->right = z;

    cout << s.isSameTree(a, x) << endl;
    int res;
    cin >> res;
}

File: 131. 分割回文串 ⭐ 深度搜索？/ac.py
class Solution:
    temp = []
    res = []
    def partition(self, s):
        
        self.f(s, 0, self.temp)
        return self.res
    

    def f(self, s, l, temp):

        if l == len(s):
            self.res.append(list(temp))
            # temp.pop(-1)
            return 

        for i in range(l, len(s)):
            t = s[l : i + 1]
            if t == t[::-1]:
                temp.append(t)
                self.f(s, i+1, temp)
                temp.pop(-1)
            else:
                continue
        
s = Solution()
print(s.partition('a'))